controllers/apps/v1beta4/add_listener.go

```diff
@@ -0,0 +1,266 @@
+package v1beta4
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	emperror "emperror.dev/errors"
+	"github.com/emqx/emqx-operator/apis/apps/v1beta4"
+	"github.com/tidwall/gjson"
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type addListener struct {
+	*EmqxReconciler
+	*portForwardAPI
+}
+
+func (a addListener) reconcile(ctx context.Context, instance v1beta4.Emqx, _ ...any) subResult {
+	if instance.GetStatus().GetCurrentStatefulSetVersion() == "" {
+		return subResult{}
+	}
+	if !shouldUpdateService(instance) {
+		return subResult{result: ctrl.Result{RequeueAfter: time.Second}}
+	}
```

Comment:
I think we don't need this check, because if EMQX cluster not ready, the portForwardAPI will be `nil`, and this function will return after `generateListenerService`